docs: add recipe for session and authentication (#27287)
nuxt/nuxt: docs/2.guide/4.recipes/4.sessions-and-authentication.md

`nuxt-auth-utils` provides a convenient `useUserSession` composable which we'll use to check if the user is logged in, and redirect them if they are not.

We'll create a middleware in the `/middleware` directory. Unlike on the server, client-side middleware is not automatically applied to all endpoints, and we'll need to specify where we want it applied.

```typescript [middleware/authenticated.ts]
export default defineNuxtRouteMiddleware(() => {
  const { loggedIn } = useUserSession()

  // redirect the user to the login screen if they're not authenticated
  if (!loggedIn.value) {
    return navigateTo('/login')
  }
})
```

## Home Page

Now that we have our app middleware to protect our routes, we can use it on our home page that display our authenticated user informations. If the user is not authenticated, they will be redirected to the login page.

> We'll use [`definePageMeta`](h/docs/api/utils/define-page-meta) to apply the middleware to the route that we want to protect.

Reviewer: ```suggestion
We'll use [`definePageMeta`](/docs/api/utils/define-page-meta) to apply the middleware to the route that we want to protect.
```

---
title: 'Sessions and Authentication'
description: "User registration and authentication is an extremely common requirement in web apps. This recipe will show you how to implement basic user registration and authentication in you Nuxt app."
---

## Introduction

In this recipe we'll be using [Drizzle](https://orm.drizzle.team/) with [db0](https://db0.unjs.io/) for database queries, but you can use any ORM or database connection strategy you prefer.

You'll need a `users` table in your database with the following columns:
- `id` (int, primary key, auto increment)
- `email` (varchar)
- `password` (varchar)

Additionally, we'll use [nuxt-aut-utils](https://github.com/Atinux/nuxt-auth-utils) by [Atinux](https://github.com/Atinux) to handle the authentication and session management.

## Steps

### 1. Install nuxt-auth-utils

```bash

> pnpm install nuxt-auth-utils

Reviewer 2: ```suggestion
npx nuxi@latest module add auth-utils
```

---
title: 'Sessions and Authentication'
description: "Authentication is an extremely common requirement in web apps. This recipe will show you how to implement basic user registration and authentication in your Nuxt app."
---

## Introduction

> In this recipe we'll be setting up user registration, login, sessions, and authentication in a full-stack Nuxt app.
>  We'll be using [Nuxt Auth Utils](https://github.com/Atinux/nuxt-auth-utils) by [Atinux (Sébastien Chopin)](https://github.com/Atinux) which provides convenient utilities for managing client-side and server-side session data. We'll install and use this to get the core session management functionality we're going to need to manage user logins. For the database ORM we'll be using [Drizzle](https://orm.drizzle.team/) with [db0](https://db0.unjs.io/), but you can use any ORM or database connection strategy you prefer.

Reviewer 2: ```suggestion
In this recipe we'll be setting up user registration, login, sessions, and authentication in a full-stack Nuxt app using [Nuxt Auth Utils](https://github.com/Atinux/nuxt-auth-utils) which provides convenient utilities for managing client-side and server-side session data. We'll install and use this to get the core session management functionality we're going to need to manage user logins. For the database ORM we'll be using [Drizzle](https://orm.drizzle.team/) with the built-in [Nitro SQL database](https://nitro.unjs.io/guide/database), but you can use any ORM or database connection strategy you prefer.
```

---
title: 'Sessions and Authentication'
description: "User registration and authentication is an extremely common requirement in web apps. This recipe will show you how to implement basic user registration and authentication in you Nuxt app."
---

## Introduction

In this recipe we'll be setting up user registration, login, sessions, and authentication in a full-stack Nuxt app.

>  We'll be using [Nuxt Auth Utils](https://github.com/Atinux/nuxt-auth-utils) by [Altinux (Sébastien Chopin)](https://github.com/Atinux) which provides convenient utilities for managing front-end and back-end session data. We'll install and use this to get the core session management functionality we're going to need to manage user logins. For the database ORM we'll be using [Drizzle](https://orm.drizzle.team/) with [db0](https://db0.unjs.io/), but you can use any ORM or database connection strategy you prefer.

Reviewer 3: I think you mean:
```diff
- [Altinux (Sébastien Chopin)]
+ [Atinux (Sébastien Chopin)]
```

You'll need a `users` table in your database with the following columns:
- `id` (int, primary key, auto increment)
- `email` (varchar)
- `password` (varchar)

> Additionally, we'll use [nuxt-aut-utils](https://github.com/Atinux/nuxt-auth-utils) by [Atinux](https://github.com/Atinux) to handle the authentication and session management.

Reviewer 3: Considering this was already specified in the introduction, it may not need to be repeated again.
Also it makes for inconsistent capitalization and syntax with the one in introduction as well as a typo (aut in stead of auth)

Author: Good feedback. Thanks.

## Steps

### 1. Install nuxt-auth-utils

> Install the [auth-utils](https://github.com/Atinux/nuxt-auth-utils) module using the `nuxi` CLI.

Reviewer 3: Likely no need to repeat this again here, we can probably say:

```md
Install the module using the `nuxi` CLI.
```

Author: I think this is important to leave in. This is the first step, and the first action that a person needs to actually perform. People skip introductions all the time, so I wouldn't count on that being a place for instructions.

Reviewer 3: Although I personally believe it would be the reader's fault for skipping chunks of information, if you think it's worth keeping in, we should probably keep it consistent with the other references, so Nuxt Auth Utils.
Though realistically speaking, there's only so much we can do to aid those who skip documentation without hurting those who don't

```bash
npx nuxi@latest module add auth-utils
```

### 1a. (Optional) Add a session encryption key

> Session cookies are encrypted. The encryption key is set from the `.env` file. This key will be added to your `.env` automatically when running in development mode the first time. However, you'll need to add this to your production environment before deploying.

Reviewer 3: This sentence seems a bit too.. "hard", there are a lot of stops here, might be worth rewording into something like:
```md
Session cookies are encrypted using a key set from the `.env` file. This key will be added to your `.env` automatically when running in development mode the first time.
```

Author: Updated. This is very good feedback on readability and flow.

Reviewer 3: Glad I can help!

Reviewer 3: @Smef 
Actually, just about about the `set` that is also redundant, as the next sentence talks about the key setter, so it can stay 
```md
using a key from the `.env` file.
```

Author: Updates

```dotenv [.env]
NUXT_SESSION_PASSWORD=password-with-at-least-32-characters
```

### 2. Create a registration page

The first page we'll need is a page for users to register and create new accounts. Create a new Vue page in your Nuxt app at `/pages/register.vue` for user registration. This page should have a form with fields for email and password. We'll intercept the form submission using `@submit.prevent` and use the [`$fetch`](/docs/getting-started/data-fetching#fetch) utility to post the data to `/api/register`. This form POST will be received by Nuxt in an API route which we will set up next. 

> If the request is successful, we'll navigate to the (soon to be created) `/users` page, which will be guarded and only be visible to logged in users.

Reviewer 3: Likely worth rewording into something like "which will be guarded and only visible to logged in users.". The second `be` is a bit redundant

Here's an example registration form for reference:

```vue [pages/register.vue]
<script setup lang="ts">
  const form = ref({
    name: "",
    email: "",
    password: "",
  });

  const error = ref(null);

  async function submitForm() {
    // clear any previous errors
    error.value = null;

    // perform the login
    try {
      await $fetch("/api/register", {
        method: "POST",
        body: form.value,
      });
    } catch (e) {
      // if there's an error, set the error message and return early
      error.value = "Error Registering User";
      return;
    }
    // Refresh the session status now that the user is logged in
    // This is a fetch() function from auth-utils to refresh session data in the client, and is not the same as $fetch
    const { fetch } = useUserSession();
    await fetch();
    // you may want to use something like Pinia to manage global state of the logged-in user
    // update Pinia state here...

>     // take the user to the auth-only users index page now that they're logged in

Reviewer 3: "the auth-only users index page" sounds odd. Perhaps it can be rephrased into "the authorized users page"?

    // window.location.href = "/users";
  }
</script>

<template>
  <div>
    <form class="space-y-2" @submit.prevent="submitForm">
      <div>
        <label for="name" class="block text-sm uppercase">Name</label>

        <input id="name" name="name" v-model="form.name" class="rounded-md px-2 py-1" required />
      </div>

      <div>
        <label for="email" class="block text-sm uppercase">Email</label>

        <input id="email" name="email" v-model="form.email" class="rounded-md px-2 py-1" required />
      </div>

      <div>
        <label for="password" class="block text-sm uppercase">password</label>

        <input
            id="password"
            name="password"
            v-model="form.password"
            class="rounded-md px-2 py-1"
            type="password"
            required
        />
      </div>

      <div class="pt-2">
        <button type="submit">Register</button>
      </div>

      <div v-if="error">{{ error }}</div>
    </form>
  </div>
</template>

```

### 3. Create an API route for registration

> With the user interface created we'll need to add a route to receive the registration form data. This route should accept a POST request with the email and password in the request body. It should hash the password and insert the user into the database. This route will only accept POST requests, so we'll follow the instructions for [API route methods](https://nuxt.com/docs/guide/directory-structure/server#matching-http-method) and name the file with `*.post.ts` to restrict the endpoint to only accept POST.

Reviewer 3: "file with `*.post.ts`" can likely be rephrased into "file `*.post.ts`", as `name the file with` doesn't quite fit (you name something some name, you don't name something with some name)

"to restrict the endpoint to only accept POST." seems a little convoluted and can likely be rephrased into "to restrict the endpoint to POST requests".

After we've successfully registered the user and stored the record in the database we can log them in by calling the `replaceUserSession` utility function from auth-utils. This utility function is automatically imported by the auth-utils module. We're using `replaceUserSession` here to make sure that any existing session data is cleared and replaced with the user login we're performing now.

The example file below is a very simple example of registration. You would probably want to add some error handling and nice response messages.

```typescript [/server/api/register.post.ts]
import users from "~/database/schema/users";

> import getDatabase from "~/database/database";

Reviewer 3: The database should be stored in ~/server to ensure it is in the server build.

import bcrypt from "bcrypt";

export default defineEventHandler(async (event) => {
  const body = await readBody(event);

  const db = await getDatabase();

>   // has the password before creating the user record

Reviewer 3: ```diff
-   // has the password before creating the user record
+   // hash the password before creating the user record
```

This is a very simple login form example, so you'd definitely want to add more validation and error checking in a real-world application.

```vue [/pages/login.vue]
<script setup lang="ts">
  import { ref } from "vue";

  const form = ref({
    email: "",
    password: "",
  });

  const error = ref(null);

  async function submitForm() {
    // clear any previous errors
    error.value = null;

    // perform the login
    try {
      await $fetch("/api/auth/login", {
        method: "POST",
        body: loginForm.value,
      });
    } catch (e) {
      // if there's an error, set the error message and return early
      error.value = "Error logging in";
      return;
    }

    // Refresh the session status now that the user is logged in
    // This is a fetch() function from auth-utils to refresh session data in the client, and is not the same as $fetch
    const { fetch } = useUserSession();
    await fetch();
    
    // you may want to use something like Pinia to manage global state of the logged-in user
    // update Pinia state here...

>     // take the user to the auth-only users index page now that they're logged in

Reviewer 3: Same suggestion as the one in the signup form

      </div>

      <div>
        <label for="password" class="block text-sm uppercase">password</label>

        <input
            id="password"
            name="password"
            v-model="form.password"
            class="rounded-md px-2 py-1"
            type="password"
            required
        />
      </div>

      <div class="pt-2">
        <button type="submit">Login</button>
      </div>

      <div v-if="error">{{ error }}</div>
    </form>
  </div>
</template>

```

### 5. Create an API route for login

With the login form created, we need to create an API route to handle the login request. This route should accept a POST request with the email and password in the request body and check the email and password against the database. If the user and password match, we'll set a session cookie to log the user in.

This server API route should be at `/server/api/auth/login.post.ts`. Just like the registration form endpoint, suffixing the filename in `.post.ts.` means that this handler will only respond to post requests.

```typescript [/server/api/auth/login.post.ts]
import users from "~/database/schema/users";

> import getDatabase from "~/database/database";

Reviewer 3: Here too, this should be moved to ~/server/

import { eq } from "drizzle-orm";
import bcrypt from "bcrypt";

export default defineEventHandler(async (event) => {
  const db = await getDatabase();

  const foundUser = (
    await db
      .select({ id: users.id, name: users.name, email: users.email, password: users.password })
      .from(users)
      .where(eq(users.email, email))
      .limit(1)
  )?.[0];

  // compare the password hash
  if (!foundUser || !bcrypt.compareSync(password, foundUser.password)) {
    // return an error if the user is not found or the password doesn't match
    throw createError({
      statusCode: 401,
      statusMessage: "Invalid email or password",
    });
  }

  // log in as the selected user
  await replaceUserSession(event, {
    user: {
      id: user.id,
      name: user.name,
    },
    loggedInAt: new Date(),
  });
});

```

> The user should now be logged in! With the session set, we can get the current user session in any API route or page by calling `getUserSession(event)` which is auto-imported as a util function from the `nuxt-auth-utils` package.

Reviewer 3: "by calling" can likely be rephrased into "using"
"auto-imported as a util function from the `nuxt-auth-utils` package." can likely be rephrased into "auto-imported from the module", as there's no need to repeat the name of the package too much =)

### 6. Create a logout API route

> Users need to be able to log out, so we should create an API route to allow them to do this. This should require post request as well, just like login. We'll clear the session when this endpoint is called.

Reviewer 3: "to do so" would likely sound better than "to do this"
Also, "This should require post request" should add an `a` (a POST request)

> We'll use the `clearUserSession` from the `auth-utils` module to log the user out and clear the session data. This function is automatically imported from the module by Nuxt, so we don't need to manually import it.

Reviewer 3: We can also remove auth-utils here and reword it into 
```md
We'll use the `clearUserSession` function to log...
```
There's also a question to be had whether it should also specify the auto-import, as this can also get redundant pretty quickly.

```typescript [server/api/auth/logout.post.ts]
export default defineEventHandler(async (event) => {
  // Clear the current user session
  await clearUserSession(event);
});
```

### 7. Protect your server route

> Protecting server routes is key to making sure your data are safe. Front-end middleware is helpful for the user, but without back-end protection your data can still be accessed. Because of this, it is critical that we protect any API routes with sensitive data. For these sensitive routes, we should return a 401 error if the user is not logged in.

Reviewer 3: data is safe (data is singular in this specific case)
Also, to use less technical terms, we might want to rephrase front-end and back-end into client-side and server-side.

> The `auth-utils` module provides the `requireUserSession` utility function to help make sure that users are logged in and have an active session. We can use this to protect our different endpoints. Like many of the other utilities from the auth module, it is automatically imported in our server endpoints.

Reviewer 3: ```md
The `requireUserSession` utility function can help make sure...
```

> In the example below, we use the `requireUserSession` utility function to protect the `/server/api/users.get.ts` server route. This route will only be accessible to logged-in users.

Reviewer 3: "In the example below, we use it to protect..."

```typescript [/server/api/users.get.ts]
import getDatabase from "~/database/database";

> import users from "~/database/schema/users";

Reviewer 3: And here three, this should be moved to ~/server/

> import requireUserLoggedIn from "~/server/utils/requireUserLoggedIn";

Reviewer 3: This import is not used and could just confuse people

export default defineEventHandler(async (event) => {
  // make sure the user is logged in
  // This will throw a 401 error if the request doesn't come from a valid user session
  await requireUserSession(event);

>   // If we make it hear, the user is authenticated. It's safe to fetch and return data

Reviewer 3: If we make it here

Author: Ha. Fixed.

  const db = await getDatabase();
  
  // Send back the list of users
  const userList = await db.select({ name: users.name, id: users.id }).from(users).limit(10);

  return userList;
});
```

> ### 8. Create a front-end middleware to protect routes

Reviewer 3: Here too we should consider using the more friendly client-side term in favor of front-end

Author: I've replaced "front-end" with "client-side" and "back-end" with "server-side" throughout the document.

Reviewer 3: Sounds good! The less technical terms the more beginner friendly it would be

> Our data are safe with the back-end route in place, but without doing anything else, unauthenticated users would probably get some odd data when trying to access the `/users` page. We should create a [front-end middleware](https://nuxt.com/docs/guide/directory-structure/middleware) to protect the route on the client side and redirect users to a login page.

Reviewer 3: Data is safe here as well, and the same suggestion about back-end and front-end

"redirect users to a login page." should use `the` in stead of `a`, as there is only one specific login page:
"redirect users to the login page."

Reviewer 3: I agree that grammatically it is correct, but the nuance is about finding the right amount of compromise between grammar and beginner-friendliness, similar to front-end and back-env vs client-side and server-side. Though all in all, I believe this could stay as is as it doesn't change much

Author: Alright. I wanted to make sure this was taken into consideration. I've pushed up the language change since you think it'll be easier for readers.

Reviewer 3: Sounds good to me! Even if you do choose to revert it in the end it would be totally fine, as this has very little impact on the recipe after all.

> `nuxt-auth-utils` provides a convenient `useUserSession` composable which we'll use to check if the user is actually logged in, and redirect them if they are not.

Reviewer 3: "actually" seems a bit out-of-place here, and can safely be omitted

We'll create a middleware in the `/middleware` directory. Unlike on the server, front-end middleware is not automatically applied to all endpoints, and we'll need to specify where we want it applied.

```typescript [/middleware/RedirectIfNotAuthenticated.ts]
export default defineNuxtRouteMiddleware(() => {
    // check if the user is logged in
    const { loggedIn } = useUserSession();

    // redirect the user to the login screen if they're not authenticated
    if (!loggedIn.value) {
        return navigateTo("/login");
    }

    return null;
});
```

### 9. Protect a route with the front-end middleware

Now that we have the front-end middleware to protect front-end routes, we can use it in any page to ensure that only logged-in users can access the route. Users will be redirected to the login page if they are not authenticated.

We'll use [`definePageMeta`](https://nuxt.com/docs/api/utils/define-page-meta) to apply the middleware to the route that we want to protect.

::important

> :warning: Remember that your data aren't really secure without back-end protection! Always secure your data on the back-end first before worrying about the front-end.

Reviewer 3: "Data isn't really" 😄 
Also same suggestions about back-end and front-end

Author: Fixed, back/front, left is/are for now.
